### ISSUE TITLE
dynamic declared variables in xquery module lead to NPE

### ISSUE THREAD

**Issue by line-o** [eXist-db/exist#2437]
### What is the problem

It is possible to declare variables, whose values will be (lazily?) evaluated at runtime.
Using a function or arrow expression leads to an NPE.

### What did you expect

one of the three:

1. the value to be evaluated correctly
2. the xquery parser to raise an error
3. the xquery runtime to raise an error

The xquery 3.1 spec states:

> [Definition: If a variable declaration includes an expression (VarValue or VarDefaultValue), the expression is called an initializing expression. The static context for an initializing expression includes all functions, variables, and namespaces that are declared or imported anywhere in the Prolog, other than the variable being declared.]

and includes following example

```xquery
declare variable $a := local:f(); 
declare variable $b := 1;
declare function local:f() { $b }; 
```

### Describe how to reproduce or add a test

Run the following test suite:

```xquery
xquery version "3.1";

module namespace devar="http://exist-db.org/xquery/test/declared-variables";
declare namespace test="http://exist-db.org/xquery/xqsuite";

declare variable $devar:static := 'I am static.';
declare variable $devar:dynamic := replace($devar:static, 'static', 'dynamic');
declare variable $devar:function := (function ($a) {
    replace($a, 'static', 'the return of a function expression')
})($devar:static);

declare variable $devar:arrow := $devar:static 
    => replace('static', 'the return of an arrow expression')
;

declare
    %test:assertEquals('I am static.')
function devar:read-static () {
    $devar:static
};

declare
    %test:assertEquals('I am dynamic.')
function devar:read-dynamic () {
    $devar:dynamic
};

declare
    %test:assertEquals('I am the return of a function expression.')
function devar:read-function () {
    $devar:function
};

declare
    %test:assertEquals('I am the return of a arrow expression.')
function devar:read-arrow () {
    $devar:arrow
};
```

### Context information
Please always add the following information

- eXist-db version 5.0.0 / 3d8272ad5fb706cb6291cabf8d1e3cef31566453 
- Java(TM) SE Runtime Environment (build 1.8.0_152-b16)
- Operating system MacOs Mojave
- How is eXist-db installed? GitHub
- Any custom changes in e.g. conf.xml none

**Comment by joewiz** [comment]
I can confirm the result described here. The test suite returns the following output:

```xml
<testsuites>
    <testsuite package="http://exist-db.org/xquery/test/declared-variables"
        timestamp="2019-01-25T13:51:44.605-05:00" tests="4" failures="0" errors="2" pending="0"
        time="PT0.009S">
        <testcase name="read-arrow" class="devar:read-arrow">
            <error type="java:java.lang.NullPointerException" message=""/>
        </testcase>
        <testcase name="read-dynamic" class="devar:read-dynamic"/>
        <testcase name="read-function" class="devar:read-function">
            <error type="java:java.lang.NullPointerException" message=""/>
        </testcase>
        <testcase name="read-static" class="devar:read-static"/>
    </testsuite>
</testsuites>
```

Unfortunately exist.log contains no trace of the NPEs... just the following warning:

> 2019-01-25 13:53:02,357 [qtp144796901-336] WARN  (PathExpr.java [replace]:119) - Expression not found: $devar:static; in: devar:read-static() item()*

**Comment by line-o** [comment]
In my real-world code I get the following stacktrace for the arrow expression:
```
java.lang.NullPointerException
	at org.exist.xquery.ArrowOperator.eval(ArrowOperator.java:100)
	at org.exist.xquery.AbstractExpression.eval(AbstractExpression.java:71)
	at org.exist.xquery.ArrowOperator.eval(ArrowOperator.java:94)
	at org.exist.xquery.AbstractExpression.eval(AbstractExpression.java:71)
	at org.exist.xquery.PathExpr.eval(PathExpr.java:276)
	at org.exist.xquery.DynamicCardinalityCheck.eval(DynamicCardinalityCheck.java:74)
	at org.exist.xquery.Atomize.eval(Atomize.java:67)
	at org.exist.xquery.ConcatExpr.eval(ConcatExpr.java:44)
	at org.exist.xquery.AbstractExpression.eval(AbstractExpression.java:71)
	at org.exist.xquery.PathExpr.eval(PathExpr.java:276)
	at org.exist.xquery.VariableDeclaration.eval(VariableDeclaration.java:123)
	at org.exist.xquery.AbstractExpression.eval(AbstractExpression.java:71)
	at org.exist.xquery.ExternalModuleImpl.resolveVariable(ExternalModuleImpl.java:226)
	at org.exist.xquery.ModuleContext.resolveVariable(ModuleContext.java:388)
	at org.exist.xquery.VariableReference.getVariable(VariableReference.java:102)
	at org.exist.xquery.PathExpr.eval(PathExpr.java:218)
	at org.exist.xquery.DynamicCardinalityCheck.eval(DynamicCardinalityCheck.java:74)
	at org.exist.xquery.Atomize.eval(Atomize.java:67)
	at org.exist.xquery.ConcatExpr.eval(ConcatExpr.java:44)
	at org.exist.xquery.functions.fn.FunDoc.eval(FunDoc.java:94)
	at org.exist.xquery.InternalFunctionCall.eval(InternalFunctionCall.java:41)
	at org.exist.xquery.AbstractExpression.eval(AbstractExpression.java:71)
	at org.exist.xquery.PathExpr.eval(PathExpr.java:276)
	at org.exist.xquery.VariableDeclaration.eval(VariableDeclaration.java:123)
	at org.exist.xquery.AbstractExpression.eval(AbstractExpression.java:71)
	at org.exist.xquery.ExternalModuleImpl.resolveVariable(ExternalModuleImpl.java:226)
	at org.exist.xquery.ModuleContext.resolveVariable(ModuleContext.java:388)
	at org.exist.xquery.VariableReference.getVariable(VariableReference.java:102)
	at org.exist.xquery.PathExpr.eval(PathExpr.java:218)
	at org.exist.xquery.VariableDeclaration.eval(VariableDeclaration.java:123)
	at org.exist.xquery.AbstractExpression.eval(AbstractExpression.java:71)
	at org.exist.xquery.ExternalModuleImpl.resolveVariable(ExternalModuleImpl.java:226)
	at org.exist.xquery.ModuleContext.resolveVariable(ModuleContext.java:388)
	at org.exist.xquery.VariableReference.returnsType(VariableReference.java:132)
	at org.exist.xquery.PathExpr.returnsType(PathExpr.java:432)
	at org.exist.xquery.DynamicCardinalityCheck.returnsType(DynamicCardinalityCheck.java:116)
	at org.exist.xquery.ConcatExpr.add(ConcatExpr.java:25)
	at org.exist.xquery.parser.XQueryTreeParser.expr(XQueryTreeParser.java:366)
	at org.exist.xquery.parser.XQueryTreeParser.prolog(XQueryTreeParser.java:4710)
	at org.exist.xquery.parser.XQueryTreeParser.libraryModule(XQueryTreeParser.java:4030)
	at org.exist.xquery.parser.XQueryTreeParser.module(XQueryTreeParser.java:3861)
	at org.exist.xquery.parser.XQueryTreeParser.xpath(XQueryTreeParser.java:3642)
	at org.exist.xquery.XQueryContext.compileModule(XQueryContext.java:3080)
	at org.exist.xquery.XQueryContext.compileOrBorrowModule(XQueryContext.java:3012)
	at org.exist.xquery.XQueryContext.importModule(XQueryContext.java:2924)
	at org.exist.xquery.parser.XQueryTreeParser.importDecl(XQueryTreeParser.java:6312)
	at org.exist.xquery.parser.XQueryTreeParser.prolog(XQueryTreeParser.java:5340)
	at org.exist.xquery.parser.XQueryTreeParser.mainModule(XQueryTreeParser.java:4042)
	at org.exist.xquery.parser.XQueryTreeParser.module(XQueryTreeParser.java:3987)
	at org.exist.xquery.parser.XQueryTreeParser.xpath(XQueryTreeParser.java:3642)
	at org.exist.xquery.XQuery.compile(XQuery.java:128)
	at org.exist.xquery.XQuery.compile(XQuery.java:79)
	at org.exist.xquery.XQuery.compile(XQuery.java:71)
	at org.exist.http.servlets.XQueryServlet.process(XQueryServlet.java:442)
	at org.exist.http.servlets.XQueryServlet.doPost(XQueryServlet.java:192)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:865)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:535)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:566)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.eclipse.jetty.server.Dispatcher.forward(Dispatcher.java:163)
	at org.eclipse.jetty.server.Dispatcher.forward(Dispatcher.java:73)
	at org.exist.http.urlrewrite.Forward.doRewrite(Forward.java:51)
	at org.exist.http.urlrewrite.XQueryURLRewrite.doRewrite(XQueryURLRewrite.java:525)
	at org.exist.http.urlrewrite.XQueryURLRewrite.service(XQueryURLRewrite.java:344)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:857)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:535)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:219)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:126)
	at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:674)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:531)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:352)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:760)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:678)
	at java.lang.Thread.run(Thread.java:748)

```

**Comment by joewiz** [comment]
@line-o Ah, the xqsuite test runner must swallow the NPEs and prevent the full error from being logged... bug or feature, I wonder?

**Comment by line-o** [comment]
This issue is fixed! Tested with 5.3.0-SNAPSHOT 093c4c0726e02868acadbce9583ec5809f462b33
